# Use Case for Hackathon

### We get a dataset, we have to understand it and get value from it

A school wants to hire us to see whether we are able to understand some problems they believe they are facing. 

Something is going wrong with the performance of the students but they are not able to figure it out and give reasons to it. 

Some parents are very disappointed of the grades of their kids and have complaint continiously to the board of the school. 

Our company is low on budget and we accept the deal even if it is risky, stressful and the money won't cover much of our runaway either but we take what we get. 

## Let's get to it and try to help the School!

### Import dataset and do the first general analysis

In [1]:
import pandas as pd

In [2]:
url='https://drive.google.com/file/d/1ETCwd-hUZ_qhQX0M0OY6OAddVE4bCSOW/view?usp=sharing'
url='https://drive.google.com/uc?id=' + url.split('/')[-2]

df = pd.read_csv(url)
df.head()

gender race/ethnicity parental level of education         lunch  \
0    Male        group A                 high school      standard   
1  Female        group D            some high school  free/reduced   
2    Male        group E                some college  free/reduced   
3    Male        group B                 high school      standard   
4    Male        group E          associate's degree      standard   

  test preparation course  math score  reading score  writing score  \
0               completed          67             67             63   
1                    none          40             29             55   
2                    none          59             60             50   
3                    none          77             78             68   
4               completed          78             73             68   

   id_student  address  Year   Age  
0        1000      NaN  2023  14.0  
1        1001      NaN  2023  17.0  
2        1002      NaN  2023  14.0  
3        1003      NaN  2023  17.0  
4        1004      NaN  2023  16.0

In [3]:
df.describe()

math score  reading score  writing score   id_student  address  \
count  1010.000000    1010.000000    1010.000000  1010.000000      0.0   
mean     66.420792      64.930693      70.289109  1494.599010      NaN   
std      15.481187      18.990769      19.125831   291.537945      NaN   
min      13.000000      15.000000      23.000000  1000.000000      NaN   
25%      56.000000      54.000000      58.000000  1242.250000      NaN   
50%      66.000000      68.000000      68.000000  1494.500000      NaN   
75%      77.000000      78.000000      79.000000  1746.750000      NaN   
max     120.000000     100.000000     100.000000  1999.000000      NaN   

              Year         Age  
count  1010.000000  943.000000  
mean   2021.104950   15.483563  
std       7.681396    1.151707  
min    1990.000000   14.000000  
25%    2023.000000   14.000000  
50%    2023.000000   16.000000  
75%    2023.000000   17.000000  
max    2023.000000   17.000000

In [ ]:
# Adress, math score, year, id_student

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1010 non-null   object 
 1   race/ethnicity               1010 non-null   object 
 2   parental level of education  1010 non-null   object 
 3   lunch                        1010 non-null   object 
 4   test preparation course      1010 non-null   object 
 5   math score                   1010 non-null   int64  
 6   reading score                1010 non-null   int64  
 7   writing score                1010 non-null   int64  
 8   id_student                   1010 non-null   int64  
 9   address                      0 non-null      float64
 10  Year                         1010 non-null   int64  
 11  Age                          943 non-null    float64
dtypes: float64(2), int64(5), object(5)
memory usage: 94.8+ KB


In [ ]:
#adress no sirve para nada

In [6]:
df.shape

(1010, 12)

In [7]:
df.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score', 'id_student', 'address', 'Year', 'Age'],
      dtype='object')

In [8]:
df.isna().sum()

gender                            0
race/ethnicity                    0
parental level of education       0
lunch                             0
test preparation course           0
math score                        0
reading score                     0
writing score                     0
id_student                        0
address                        1010
Year                              0
Age                              67
dtype: int64

## What information do we get from this? 

1. Why is student id int? If it is a string. It is not a number

2. We need to inspect all the object columns and check the value counts. What do we discover when we do it? 

3. The distribution of the year column is a bit odd. We need to look more into it. 

4. Do we have duplicates in the dataset?

5. All rows of the column address are missing, maybe we can just delete it

6. The age column has 77 missing values. What should we do with it? 

7. There seems to be values for the math scores higher than 120. Are those outliers? 

### 1. Change type of student id  

In [9]:
df['id_student'] = df['id_student'].astype(str)

In [11]:
df.dtypes

gender                          object
race/ethnicity                  object
parental level of education     object
lunch                           object
test preparation course         object
math score                       int64
reading score                    int64
writing score                    int64
id_student                      object
address                        float64
Year                             int64
Age                            float64
dtype: object

### 2. Explore the columns that are objects (string)

In [15]:
for col in df.columns:
    if col != 'id_student' and df[col].dtype == 'object':
        print(df[col].value_counts())

#df['gender'].value_counts()

gender
male      373
female    337
FEMALE    103
MALE       97
Male       54
Female     46
Name: count, dtype: int64
race/ethnicity
group C    324
group D    265
group B    206
group E    134
group A     81
Name: count, dtype: int64
parental level of education
some college          225
high school           206
associate's degree    204
some high school      192
bachelor's degree     113
master's degree        70
Name: count, dtype: int64
lunch
standard        659
free/reduced    351
Name: count, dtype: int64
test preparation course
none         672
completed    338
Name: count, dtype: int64


### 3. Seems that the column gender has a lot of versions we are going to have to fix this

In [16]:
df['gender'] = df['gender'].str.upper()
df['gender'].value_counts()

gender
MALE      524
FEMALE    486
Name: count, dtype: int64

### 4. What may be happening with the year? 

In [22]:
df['Year'].value_counts()

Year
2023    942
1990     58
Name: count, dtype: int64

In [24]:
df['Year'] = df['Year'].replace(1990,2023)
df['Year'].value_counts()

Year
2023    1000
Name: count, dtype: int64

### 5. Do we have any duplicates? 

In [17]:
df.shape

(1010, 12)

In [18]:
df = df.drop_duplicates()

In [19]:
df.shape

(1000, 12)

### 6.1. Missing values: Remember the column address

In [ ]:
df.isna().sum()
df = df.drop('address', axis = 1)

### 6.2. What about the missing values of the column Age? 

### 7. What is going on with the math scores?

In [28]:
df[df['math score'] > 100]
df = df[df['math score'] <= 100]

## Do some initial plots? EDA

### What do you find interesting?  

- To me the reading score is really intriguing --> We need to really work on understanding what is going on
- The writing score is also characteristic but in principle I am not that worried about it. 

#### We know for a fact that the id_student and the year will not be relevant for any analysis as they are the same value for the whole dataset

Maybe we can delete them from now on

#### What do we care about? The grades are the main problem. Maybe they vary based on any of the other variables

I think we can detect the biggest discrepancies at 3-4 levels: 

- Parental level of education 

- Test preparation course 

- Mybe lunch or gender (I am pretty sure the lunch should not have that much impact but let's still check it)

Let's go deeper

 



### 2-variable Analysis with Lunch column 

### 2-variable Analysis with gender column 

### 2-variable Analysis with Test preparation course column 

### What is next?

- Get the story together
- Improve your code
- Get some statistics of the differences you want to show 
- Work on the plots to show the data nicely 
- Get the presentation ready 
- Show what you found to the client